File: javascript/dispatch_job.js
#!/usr/bin/env node
// dispatch_job.py — portiert nach javascript
// Quelle: python, OpenClaw@gateway1:skills/multi-nodes-utils/scripts/dispatch_job.py
// auch in: OpenClaw@gateway2:skills/multi-nodes-utils/scripts/dispatch_job.py
// Erzeugt: 2026-08-19 durch ABSTRACTIONS_MANAGER.py

/**
 * Job Dispatcher - Verteilt Jobs auf passende Nodes
 */

const fs = require('fs');
const path = require('path');
const { spawnSync } = require('child_process');

// Node-Konfiguration
const NODES = {
    "node1": { "always_available": true, "capacity": "medium", "priority": 2 },
    "node2": { "always_available": true, "capacity": "medium", "priority": 3 },
    "node3": { "always_available": false, "capacity": "medium", "priority": 4 },
    "node5": { "always_available": false, "capacity": "low", "priority": 5, "device": "Redmi Note 11S" },
    "node7": { "always_available": true, "capacity": "high", "priority": 1 }
};

class JobDispatcher {
    /** Dispatches jobs to appropriate nodes based on weight */
    
    /**
     * Bewertet Job-Gewicht
     * @param {string} scriptPath 
     * @param {number} targetLangsCount 
     * @returns {string}
     */
    getJobWeight(scriptPath, targetLangsCount = 1) {
        if (!fs.existsSync(scriptPath)) {
            return "medium";
        }
        
        const stats = fs.statSync(scriptPath);
        const scriptSize = stats.size;
        const totalWork = scriptSize * targetLangsCount;
        
        if (totalWork > 50000) {  // > 50KB
            return "heavy";
        } else if (totalWork > 10000) {  // > 10KB
            return "medium";
        } else {
            return "light";
        }
    }
    
    /**
     * Wählt besten Node basierend auf Job-Gewicht
     * @param {string} jobWeight 
     * @returns {string}
     */
    selectNode(jobWeight) {
        let preferred = [];
        
        if (jobWeight === "heavy") {
            // Schwere Jobs → Node 7 (Docker), dann Node 2, dann Node 1
            preferred = ["node7", "node2", "node1"];
        } else if (jobWeight === "medium") {
            // Mittlere Jobs → Stable Nodes
            preferred = ["node2", "node1", "node7"];
        } else {  // light
            // Leichte Jobs → Mobile/verfügbare Nodes
            preferred = ["node5", "node1", "node2"];
        }
        
        // Prüfe Verfügbarkeit
        for (const nodeId of preferred) {
            if (this.checkNodeAvailable(nodeId)) {
                return nodeId;
            }
        }
        
        // Fallback
        return "node1";
    }
    
    /**
     * Prüft ob Node erreichbar ist
     * @param {string} nodeId 
     * @returns {boolean}
     */
    checkNodeAvailable(nodeId) {
        if (!NODES[nodeId]) {
            return false;
        }
        
        const node = NODES[nodeId];
        
        // Nicht immer-verfügbare Nodes nur wenn explizit requested
        if (!node.always_available) {
            // Für light-jobs prüfen wir ob online
            if (nodeId === "node5") {  // Redmi
                return this._checkMobileOnline();
            }
            return false;
        }
        
        // Für immer-verfügbare Nodes: prüfe ob wirklich online
        try {
            const result = spawnSync("openclaw", ["nodes", "status", nodeId], {
                timeout: 3000
            });
            return result.status === 0;
        } catch (error) {
            return !!node.always_available;
        }
    }
    
    /**
     * Prüft ob Redmi (Node 5) Internet hat
     * @returns {boolean}
     */
    _checkMobileOnline() {
        try {
            const result = spawnSync("openclaw", ["nodes", "status", "node5"], {
                timeout: 5000
            });
            return result.status === 0 && result.stdout.toString().toLowerCase().includes("online");
        } catch (error) {
            return false;
        }
    }
    
    /**
     * Dispatched Job und gibt Info zurück
     * @param {string} jobScript 
     * @param {string[]} targetLangs 
     * @returns {Object}
     */
    dispatch(jobScript, targetLangs = null) {
        if (targetLangs === null) {
            targetLangs = ["perl5"];
        }
        
        const weight = this.getJobWeight(jobScript, targetLangs.length);
        const selectedNode = this.selectNode(weight);
        
        return {
            job: jobScript,
            weight: weight,
            selected_node: selectedNode,
            target_langs: targetLangs,
            status: "dispatched"
        };
    }
}

/**
 * Main function to parse arguments and run dispatcher
 */
function main() {
    const args = process.argv.slice(2);
    let jobPath = null;
    let targetLangsStr = "perl5";
    let forceWeight = null;
    let execute = false;
    
    // Simple argument parsing
    for (let i = 0; i < args.length; i++) {
        if (args[i] === "--job" || args[i] === "-j") {
            jobPath = args[++i];
        } else if (args[i] === "--langs" || args[i] === "-l") {
            targetLangsStr = args[++i];
        } else if (args[i] === "--weight" || args[i] === "-w") {
            forceWeight = args[++i];
        } else if (args[i] === "--execute" || args[i] === "-x") {
            execute = true;
        }
    }
    
    if (!jobPath) {
        console.error("❌ Job path is required");
        process.exit(1);
    }
    
    if (!fs.existsSync(jobPath)) {
        console.error(`❌ Job not found: ${jobPath}`);
        process.exit(1);
    }
    
    const dispatcher = new JobDispatcher();
    const targetLangs = targetLangsStr.split(",");
    
    // Determine weight
    const weight = forceWeight || dispatcher.getJobWeight(jobPath, targetLangs.length);
    
    // Select node
    const selectedNode = dispatcher.selectNode(weight);
    
    // Output
    console.log("📦 Job Dispatch Information");
    console.log("=".repeat(50));
    const stats = fs.statSync(jobPath);
    console.log(`Job: ${jobPath}`);
    console.log(`Size: ${stats.size} bytes`);
    console.log(`Target langs: ${targetLangs.join(", ")}`);
    console.log(`Job weight: ${weight}`);
    console.log(`Selected node: ${selectedNode}`);
    console.log("=".repeat(50));
    
    if (execute) {
        console.log(`\n🚀 Executing on ${selectedNode}...`);
        // TODO: Implement remote execution
        console.log("(Remote execution not yet implemented)");
    } else {
        console.log(`\n💡 To execute: ${process.argv[1]} --job ${jobPath} --execute`);
    }
}

if (require.main === module) {
    main();
}

module.exports = { JobDispatcher };

File: perl5/dispatch_job.pl
#!/usr/bin/perl
# dispatch_job.py — portiert nach perl5
# Quelle: python, OpenClaw@gateway1:skills/multi-nodes-utils/scripts/dispatch_job.py
# auch in: OpenClaw@gateway2:skills/multi-nodes-utils/scripts/dispatch_job.py
# Erzeugt: 2026-08-19 durch ABSTRACTIONS_MANAGER.py

use strict;
use warnings;
use Getopt::Long;
use File::Spec;
use Cwd 'abs_path';

# Node-Konfiguration
my %NODES = (
    "node1" => { always_available => 1, capacity => "medium", priority => 2 },
    "node2" => { always_available => 1, capacity => "medium", priority => 3 },
    "node3" => { always_available => 0, capacity => "medium", priority => 4 },
    "node5" => { always_available => 0, capacity => "low", priority => 5, device => "Redmi Note 11S" },
    "node7" => { always_available => 1, capacity => "high", priority => 1 },
);

package JobDispatcher;

sub new {
    my $class = shift;
    my $self = {};
    bless $self, $class;
    return $self;
}

# Bewertet Job-Gewicht
sub get_job_weight {
    my ($self, $script_path, $target_langs_count) = @_;
    $target_langs_count //= 1;

    unless (-e $script_path) {
        return "medium";
    }

    my $script_size = (stat($script_path))[7];
    my $total_work = $script_size * $target_langs_count;

    if ($total_work > 50000) {  # > 50KB
        return "heavy";
    } elsif ($total_work > 10000) {  # > 10KB
        return "medium";
    } else {
        return "light";
    }
}

# Wählt besten Node basierend auf Job-Gewicht
sub select_node {
    my ($self, $job_weight) = @_;

    my @preferred;
    if ($job_weight eq "heavy") {
        # Schwere Jobs → Node 7 (Docker), dann Node 2, dann Node 1
        @preferred = ("node7", "node2", "node1");
    } elsif ($job_weight eq "medium") {
        # Mittlere Jobs → Stable Nodes
        @preferred = ("node2", "node1", "node7");
    } else {  # light
        # Leichte Jobs → Mobile/verfügbare Nodes
        @preferred = ("node5", "node1", "node2");
    }

    # Prüfe Verfügbarkeit
    for my $node_id (@preferred) {
        if ($self->check_node_available($node_id)) {
            return $node_id;
        }
    }

    # Fallback
    return "node1";
}

# Prüft ob Node erreichbar ist
sub check_node_available {
    my ($self, $node_id) = @_;

    return 0 unless exists $NODES{$node_id};

    my $node = $NODES{$node_id};

    # Nicht immer-verfügbare Nodes nur wenn explizit requested
    if (!$node->{always_available}) {
        # Für light-jobs prüfen wir ob online
        if ($node_id eq "node5") {  # Redmi
            return $self->_check_mobile_online();
        }
        return 0;
    }

    # Für immer-verfügbare Nodes: prüfe ob wirklich online
    my $result = system("openclaw nodes status $node_id");
    if ($result == 0) {
        return 1;
    } else {
        return $node->{always_available} // 0;
    }
}

# Prüft ob Redmi (Node 5) Internet hat
sub _check_mobile_online {
    my $self = shift;

    eval {
        local $SIG{ALRM} = sub { die "timeout"; };
        alarm(5);
        my $output = `openclaw nodes status node5`;
        alarm(0);
        return ($? == 0 && $output =~ /online/i);
    };
    alarm(0);
    return 0 if $@;
    return 0;
}

# Dispatched Job und gibt Info zurück
sub dispatch {
    my ($self, $job_script, $target_langs) = @_;
    $target_langs //= ["perl5"];

    my $weight = $self->get_job_weight($job_script, scalar @$target_langs);
    my $selected_node = $self->select_node($weight);

    return {
        job => $job_script,
        weight => $weight,
        selected_node => $selected_node,
        target_langs => $target_langs,
        status => "dispatched"
    };
}

package main;

sub main {
    my $job = "";
    my $langs = "perl5";
    my $weight = "";
    my $execute = 0;

    GetOptions(
        "job|j=s" => \$job,
        "langs|l=s" => \$langs,
        "weight|w=s" => \$weight,
        "execute|x" => \$execute
    ) or die "Falsche Optionen\n";

    unless ($job) {
        die "❌ --job/-j ist erforderlich\n";
    }

    unless (-e $job) {
        print "❌ Job nicht gefunden: $job\n";
        exit 1;
    }

    my $dispatcher = JobDispatcher->new();
    my @target_langs = split(",", $langs);

    # Determine weight
    my $job_weight;
    if ($weight) {
        $job_weight = $weight;
    } else {
        $job_weight = $dispatcher->get_job_weight($job, scalar @target_langs);
    }

    # Select node
    my $selected_node = $dispatcher->select_node($job_weight);

    # Output
    print "📦 Job Dispatch Information\n";
    print "=" x 50 . "\n";
    my $size = (stat($job))[7];
    print "Job: $job\n";
    print "Size: $size bytes\n";
    print "Target langs: " . join(", ", @target_langs) . "\n";
    print "Job weight: $job_weight\n";
    print "Selected node: $selected_node\n";
    print "=" x 50 . "\n";

    if ($execute) {
        print "\n🚀 Executing on $selected_node...\n";
        # TODO: Implement remote execution
        print "(Remote execution not yet implemented)\n";
    } else {
        my $cmd = "$0 --job $job --execute";
        print "\n💡 To execute: $cmd\n";
    }
}

main() if !caller;
